File: src/Deck.js
import { Card } from "./Card.js";

export const DECK_ACTIVE_SIZE = 5;
export const DECK_RESERVE_SIZE = 2;

/**
 * Builds and validates a player's deck: 5 chosen for the active lineup,
 * 2 for reserves, drawn from their unlocked brawler pool (15 to start,
 * growing to 65, then 115 in later updates — the pool size itself lives
 * outside this class, this just enforces deck-building rules).
 *
 * @param {import("./BrawlerTemplate.js").BrawlerTemplate[]} activeTemplates
 * @param {import("./BrawlerTemplate.js").BrawlerTemplate[]} reserveTemplates
 * @param {string} ownerId
 * @returns {{ active: Card[], reserve: Card[] }}
 */
export function buildDeck(activeTemplates, reserveTemplates, ownerId) {
  if (activeTemplates.length !== DECK_ACTIVE_SIZE) {
    throw new Error(
      `Active lineup must have exactly ${DECK_ACTIVE_SIZE} brawlers (got ${activeTemplates.length})`
    );
  }
  if (reserveTemplates.length !== DECK_RESERVE_SIZE) {
    throw new Error(
      `Reserves must have exactly ${DECK_RESERVE_SIZE} brawlers (got ${reserveTemplates.length})`
    );
  }

  const allIds = [...activeTemplates, ...reserveTemplates].map((t) => t.id);
  const uniqueIds = new Set(allIds);
  if (uniqueIds.size !== allIds.length) {
    throw new Error("Deck contains duplicate brawlers — no duplicates allowed (adjust here if that changes).");
  }

  const active = activeTemplates.map((t) => {
    const card = new Card(t, ownerId);
    card.zone = "active";
    return card;
  });
  const reserve = reserveTemplates.map((t) => {
    const card = new Card(t, ownerId);
    card.zone = "reserve";
    return card;
  });

  return { active, reserve };
}

File: game/deck.py
from .card import Card


DECK_ACTIVE_SIZE = 5
DECK_RESERVE_SIZE = 2


def build_deck(
    active_templates,
    reserve_templates,
    owner_id
):
    if len(active_templates) != DECK_ACTIVE_SIZE:
        raise ValueError(
            f"Active lineup must have exactly "
            f"{DECK_ACTIVE_SIZE} brawlers "
            f"(got {len(active_templates)})"
        )

    if len(reserve_templates) != DECK_RESERVE_SIZE:
        raise ValueError(
            f"Reserves must have exactly "
            f"{DECK_RESERVE_SIZE} brawlers "
            f"(got {len(reserve_templates)})"
        )

    all_ids = [
        template.id
        for template in (
            active_templates + reserve_templates
        )
    ]

    if len(set(all_ids)) != len(all_ids):
        raise ValueError(
            "Deck contains duplicate brawlers."
        )

    active = []

    for template in active_templates:
        card = Card(template, owner_id)
        card.zone = "active"
        active.append(card)

    reserve = []

    for template in reserve_templates:
        card = Card(template, owner_id)
        card.zone = "reserve"
        reserve.append(card)

    return {
        "active": active,
        "reserve": reserve,
    }